### PR TITLE
Fix #16072 - Clicking into input field within dropdown no longer closes the dropdown

### LINKED ISSUE #16072: Clicking into text input field within dropdown closes the dropdown
Hi,

Please follow this JSfiddle for a live example: https://jsfiddle.net/znbtb8cp/1/

**Steps to reproduce:**
1. Click on the dropdown button
2. Click on the input field

**Expected behaviour:**
The field should be focused.

**Actual behaviour:**
The dropdown closes.

The project we are working on encountered this issue, and we have a workaround, but we are ready to provide a patch and a pull request with a better/long term solution if nobody has done it yet.

### PR DESCRIPTION
**Pull request by FrenchDilettante**
Fix #16072

### CODE REVIEW

**Comment by twbs-savage** [comment]
**Tests failed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **BUSTED**

Commit: f2d61e01195ffc4fa32b32b3ab96cede6c1667f0
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/54568815

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by manudwarf** [comment]
The failure on Travis seems to be come from some timeout – do I need to do anything?

**Comment by cvrebert** [comment]
@twbs-savage Retry

**Comment by cvrebert** [comment]
Please test `<textarea>` in the testcase too.

**Comment by cvrebert** [comment]
Also, excluding the `docs/` and `dist/` file changes would be appreciated.

File: js/dropdown.js [hunk 1]
```diff
@@ -96,6 +96,8 @@
 
       if (!$parent.hasClass('open')) return
 
+      if (e && e.type === 'click' && /input|textarea/i.test(e.target.tagName) && $.contains($parent[0], e.target)) return
```
Comment (cvrebert) [inline js/dropdown.js]: Should use `==` instead of `===`, per Jacob's coding style.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: f2d61e01195ffc4fa32b32b3ab96cede6c1667f0
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/54597128

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/dropdown.js (hunk 1 above)
Comment (manudwarf) [inline js/dropdown.js]: Fixed.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 57de3cf8cf7a226a820a773dfc4f78c60cff45f0
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/54609494

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

File: js/tests/unit/dropdown.js [hunk 2]
```diff
@@ -350,4 +350,30 @@ $(function () {
...
+      .find('[data-toggle="dropdown"]')
+      .bootstrapDropdown()
+      .trigger('click')
+
+    $dropdown.find('#textField')
+      .trigger('click')
+
+    assert.ok($dropdown.parent('.btn-group').hasClass('open'), '.dropdown-menu is not visible')
+
+    $dropdown.find('#textArea')
```
Comment (cvrebert) [inline js/tests/unit/dropdown.js]: I think it would be good to explicitly close and re-open the dropdown before running this 2nd test.

Comment (manudwarf) [inline js/tests/unit/dropdown.js]: Would you prefer me to open a second test rather?

Comment (cvrebert) [inline js/tests/unit/dropdown.js]: Either way is acceptable.

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 33bd773f26b48895d9f1a3755850c5b0759071c1
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/54847032

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 164e712459577ef3378b51ffa842ce8ec376d7b1
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/54847299

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by cvrebert** [comment]
Looks pretty good to me. We'll just need to squash it into a single commit when merging it.
CC: @XhmikosR @hnrch02 for review

**Comment by manudwarf** [comment]
Do you want me to do it on a separate branch and re-open a PR?

**Comment by cvrebert** [comment]
If you know how to squash, just squash and force-push to the same branch. If not, one of us on the Core Team will take care of it.

**Comment by manudwarf** [comment]
Done :)

**Comment by twbs-savage** [comment]
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 31a908fb14dc25cadb533f916f3075f8d93d5c6f
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/55008405

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

**Comment by XhmikosR** [comment]
LGTM

(Damn this style we use is hard to read :/)

**Pull request closed without merging**